arch/arm/arm64/tls.c

```diff
@@ -0,0 +1,281 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Florian Schmidt <florian.schmidt@neclab.eu>
+ *          Cyril Soldani <cyril.soldani@uliege.be>
+ *          Simon Kuenzer <simon@unikraft.io>
+ *          Robert Kuban <robert.kuban@opensynergy.com>
+ *
+ * Copyright (c) 2019, NEC Europe Ltd., NEC Corporation. All rights reserved.
+ * Copyright (c) 2021, University of Liège. All rights reserved.
+ * Copyright (c) 2022, NEC Laboratories Europe GmbH, NEC Corporation.
+ *                     All rights reserved.
+ * Copyright (c) 2022, OpenSynergy GmbH All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef CONFIG_LIBCONTEXT_CLEAR_TBSS
+#define CONFIG_LIBCONTEXT_CLEAR_TBSS 1
+#endif /* CONFIG_LIBCONTEXT_CLEAR_TBSS */
+
+#include <string.h>
+#include <uk/print.h>
+#include <uk/hexdump.h>
+#include <uk/assert.h>
+#include <uk/essentials.h>
+#include <uk/arch/types.h>
+#include <uk/arch/tls.h>
+
+#if CONFIG_LIBUKDEBUG
+#include <uk/assert.h>
+#include <uk/print.h>
+#else /* !CONFIG_LIBUKDEBUG */
+#define UK_ASSERT(..) do {} while (0)
+#define uk_pr_debug(..) do {} while (0)
+#endif /* !CONFIG_LIBUKDEBUG */
+
+/*
+ * aarch64 ABI requires 16 bytes of TCB
+ * after the TLS arch pointer
+ *
+ * - in musl that included in the size of the TCB (ptread)
+ * - other libc might need to add 16 byte after the TCB
```

Comment:
I'm not sure I understand these two sentences. @kubanrob , could you assist?

Reply:
It is explained here very well: source: https://www.openwall.com/lists/musl/2018/02/09/7

Basically: aarch64 ABI demands a block of 16 bytes after the tlsp, which demands values to be stored there. However, do to the structure of the TLS these values contain not enough to access the TLS and so the compiler generated code never accesses this block. Knowing that, musl just reuses that space for other members in the TCB.

Basically these 16bytes are part of the TCB in this layout. If you don't want them to be for some libc, you need to add 16bytes to the actual size of the TCB.

Reply:
I changed the whole block to:
```c
/*
 * Aarch64 ABI requires 16 bytes of space after the tlsp as part of the
 * TCB.
 * NOTE: We assume that libraries (like musl) reserve this space when using
 *       this TLS API. You can find more information here:
 *       https://www.openwall.com/lists/musl/2018/02/09/7
 */
```